dialogue-core/src/main/java/com/palantir/dialogue/core/QueuedChannel.java

```diff
@@ -0,0 +1,264 @@
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Scheduled {} requests", SafeArg.of("numScheduled", numScheduled));
+        }
+    }
+
+    /**
+     * Get the next call and attempt to execute it. If it is runnable, wire up the underlying future to the one
+     * previously returned to the caller. If it is not runnable, add it back into the queue. Returns true if more
+     * tasks may be able to be scheduled, and false otherwise.
+     */
+    private boolean scheduleNextTask() {
+        DeferredCall queueHead = queuedCalls.poll();
+        if (queueHead == null) {
+            return false;
+        }
+        SettableFuture<Response> queuedResponse = queueHead.response();
+        // If the future has been completed (most likely via cancel) the call should not be queued.
+        // There's a race where cancel may be invoked between this check and execution, but the scheduled
+        // request will be quickly cancelled in that case.
+        if (queuedResponse.isDone()) {
+            queueSizeEstimate.decrementAndGet();
+            return true;
+        }
+        try (CloseableSpan ignored = queueHead.span().childSpan("Dialogue-request-scheduled")) {
+            Endpoint endpoint = queueHead.endpoint();
+            Optional<ListenableFuture<Response>> maybeResponse = delegate.maybeExecute(endpoint, queueHead.request());
+
+            if (maybeResponse.isPresent()) {
+                queueSizeEstimate.decrementAndGet();
+                ListenableFuture<Response> response = maybeResponse.get();
+                queueHead.span().complete();
+                numRunningRequests.incrementAndGet();
+                response.addListener(numRunningRequests::decrementAndGet, DIRECT);
+                Futures.addCallback(response, new ForwardAndSchedule(queuedResponse), DIRECT);
+                queuedResponse.addListener(
+                        () -> {
+                            if (queuedResponse.isCancelled()) {
+                                // TODO(ckozak): Consider capturing the argument value provided to cancel to propagate
+                                // here.
+                                // Currently cancel(false) will be converted to cancel(true)
+                                if (!response.cancel(true) && log.isDebugEnabled()) {
+                                    log.debug(
+                                            "Failed to cancel delegate response, it should be reported by"
+                                                    + " ForwardAndSchedule logging",
+                                            SafeArg.of("service", endpoint.serviceName()),
+                                            SafeArg.of("endpoint", endpoint.endpointName()));
+                                }
+                            }
+                        },
+                        DIRECT);
+                return true;
+            } else {
+                if (!queuedCalls.offerFirst(queueHead)) {
```

Comment:
could we `addFirst` instead?

Reply:
addFirst can throw on failure, I’d prefer not to introduce a scenario that forces us to throw and catch. The current implementation is expected not to throw because we don’t use a queue with a maximum size, but I’d like to keep the queue safe to allow future changes (around removing elements upon cancelation) without risking reintroducing bugs.